[푸코] Step02 - 카드 클래스 구현하기

wnsxor1993 opened the pull request:
## 작업 목록
- [x] Card 클래스 설계
- [x] 프로퍼티로 사용될 Shape, Number의 데이터 구조 선택 및 설계
- [x] 프로퍼티 직접 접근 대신 메서드를 통한 접근 구현
- [x] VC 내에서 Card 인스턴스 생성 메서드 구현
- [x] VC 내에서 Card 인스턴스 정보 출력 메서드 구현

## 학습 키워드
1) 객체지향 (Class)
2) enum
3) 데이터 구조 선택 이유

## 고민과 해결
- Card 인스턴스 UIImageView 연결
: 생성된 ImageView를 클릭 시, Card 인스턴스 정보가 출력되는 형태를 구현하고 싶었으나 둘의 관계를 이어주는 방법이나 터치 이벤트 연결 방법에 대한 공부가 아직 부족하고, 대대적인 수정이 필요하여 지금 단계에서는 ImageView 생성 시, Card 인스턴스도 생성하고 출력하는 형태로 구현함.

 - 자료구조 선택
 : 이전에는 자료구조를 선택하는 이유에 대해서 크게 생각하지 않았으나 왜 해당 자료구조를 선택한 지에 대한 고민을 하고 타당한 이유를 들어야대는 부분이 다소 어려웠음. 각 자료구조 간의 차이를 알아도 아직 체감되는 부분이 적다보니 더더욱 선택 시 많은 어려움이 있는 것 같다는 생각이 듬.

wnsxor1993 committed:
Feat: Card 클래스 생성
- 카드 문양과 숫자 정보를 담고 있는 각각의 데이터 구조 구현
- 카드 정보를 리턴하는 함수까지 구현된 Card 클래스 작성

wnsxor1993 committed:
Feat: Card 인스턴스 생성 및 출력 함수 구현
- 랜덤 값을 프로퍼티에 할당하여 Card 인스턴스 생성하는 함수
- 생성된 Card 인스턴스의 값을 출력하는 함수

godrm commented on PockerGameApp/PockerGameApp/Data/Card.swift:
```diff
@@ -0,0 +1,26 @@
+// 하나의 카드가 가지게 될 Card 인스턴스는 항상 고유한 값이 되어야 하므로 class로 설계
+class Card{
+    private let number: CardNumber
+    private let shape: Shape
```
CardNumber나 Shape가 Card와 밀첩한 타입이라면 Card.Number 나 Card.Shape 형태로 nested 형태도 고려해보세요

wnsxor1993 replied:
Card 클래스와의 연관성이 높으며, 차후 개별 재사용 가능성이 낮아 두 타입 모두 nested 형태로 수정하였습니다

godrm commented on PockerGameApp/PockerGameApp/Data/Card.swift:
```diff
@@ -0,0 +1,26 @@
+    func cardInformation() -> String{
+        let cardValue = shape.rawValue + number.changedNumber
+        return cardValue
+    }
```
Swift에서는 속성을 문자열 값으로 전달하는 경우는 description을 사용합니다
CustomStringConvertible 프로토콜을 확인해보세요

wnsxor1993 replied:
커스텀 속성들과 관련하여 문자열 전달 구문들은 description 사용 형태로 수정하였습니다.

godrm replied:
문자열을 합칠 때 +보다는 `"\(shape)"` 같은 Interpolation 방식을 권장합니다. 
특히 description을 붙이지 않아도 description에서 값을 가져와서 처리합니다.

godrm commented on PockerGameApp/PockerGameApp/Data/ShapeNumber.swift:
```diff
@@ -0,0 +1,39 @@
+    }
+    
+    init(num: Int){
+        number = num
```
이렇게 선언하면 Int라서 0, 100, 1234 같은 값도 CardNumber가 가능한가요?

wnsxor1993 replied:
예외 요소를 생각하지 못 하였습니다. 이번 수정 때, guard 구문을 기입하여 예외에 대한 처리를 추가하였습니다.

godrm commented on PockerGameApp/PockerGameApp/Data/ShapeNumber.swift:
```diff
@@ -0,0 +1,39 @@
+    case heart = "♥️"
+    case dia = "♦️"
+    case spade = "♠️"
+    case clover = "♣️"
```
위에 언급한 Card와 함께 CustomStringConvertible을 고려해보세요

godrm commented on PockerGameApp/PockerGameApp/ViewController.swift:
```diff
@@ -9,31 +9,44 @@ import UIKit
+    
+    func makeCardInfo() -> Card{
+        let shapes: [Shape] = [.clover, .dia, .heart, .spade]
+        let shapeRandomNum = Int.random(in: 0...3)
```
Shape 타입을 활용한다면 랜덤으로 뽑을 때도 0...3 대신 Shape 속성중에 하나를 뽑도록 하는게 어떨까요?

wnsxor1993 replied:
이번 수정 때, CaseIterable을 채택하여 allCases를 활용한 random 요소를 구현하였습니다!

godrm reviewed (changes requested):
Card와 속성 관련해서는 수정해서 다시 Push해보세요

wnsxor1993 committed:
Chore: Card 클래스 구문 수정
- 연관 속성 Nested 형태로 수정
- String 값 출력, CustomStringConvertible 채택을 통한 description 활용
- guard 구문을 통한 예외 요소 배제
- 초기화 구문 수정

godrm commented on PockerGameApp/PockerGameApp/Data/Card.swift:
```diff
@@ -8,18 +8,71 @@
+        let cardRandomNum = Int.random(in: 1...13)
+        
+        self.number = CardNumber(num: cardRandomNum)
+        self.shape = shapeRandomNum
```
하나의 카드를 생성할 때 랜덤으로 생성하면 원하는 카드가 나올때까지 만들어야 하는걸까요?
init()할 때 Number와 Shape를 지정해주는 것에서 왜 이렇게 바뀐걸까요?

wnsxor1993 replied:
현 단계에서는 꼭 지정 값을 가진 카드를 생성할 필요는 없어보여서 랜덤 값으로 카드를 생성하도록 구현했습니다. 이후 단계에서 지정 값으로 카드가 생성되는 이니셜라이저를 추가할 생각이었습니다!ㅎㅎ

godrm replied:
생성자에 대해 오해하고 있는 것 같아요. 랜덤하게 카드를 생성하는 것은 Card.init()의 역할이 아니라 Card.init()에 값을 전달해줄 상위의 역할입니다. 이것을 구분하는 것은 중요한 포인트입니다. 이렇게 생성하는 init() 함수는 정말 써먹을 수가 없습니다
이게 지금 단계에서 필요없다고 생각하지마세요. 항상 저장소에 푸시한건 다른 사람이 가져다 써야한다고 생각하셔야 합니다. 
이렇게 만든 Card 타입을 다른 사람이 전혀 못 쓸꺼 같네요

wnsxor1993 replied:
생성자를 여러개 생성할 수 있는 부분에 생각이 팔려서 잘못된 방향으로 접근한 것 같습니다.
해당 요소를 다시 신경쓰면서 수정하도록 하겠습니다!

godrm commented on PockerGameApp/PockerGameApp/Data/Card.swift:
```diff
@@ -8,18 +8,71 @@
+        init(num: Int){
+            guard num <= 13 || num != 0 else{
+                print("적절하지 못한 값이 발견되었습니다 : \(num)")
+                fatalError()
```
fatalError는 앱 전체가 멈추는 위험한 표현입니다. 가능하면 안쓰는게 좋아요. 
더구나 에러 상황이라고 해도 앱이 종료되기보다는 사용자가 다시 선택할 수 있도록 하는 게 좋습니다. 
수업에서 언급한 것처럼 로직에서 print()로 뭔가 출력하는 것은 이제 지양하세요 

덧붙여서 CardNumber를 생성하는 측면에서도 num 이라는 숫자를 입력 받는 거라면,
failable init에 대해 학습해보길 추천드립니다. 그리고 Number도 어차피 범위가 정해진 거라면 enum을 해도 되지는 않을까요?

wnsxor1993 replied:
넵 감사합니다!ㅎㅎ 점점 수정하다보니 Struct로 타입을 유지할 이유가 없어지는 듯 합니다. 해당 타입도 다음 단계 때, enum 형태로 수정해보도록 하겠습니다.
추가적으로 말씀하신 failable init에 대해서도 다시 학습하도록 하겠습니다^^

godrm commented on PockerGameApp/PockerGameApp/ViewController.swift:
```diff
@@ -37,16 +37,12 @@ class ViewController: UIViewController {
     
     func showCardInfo(card: Card){
-        print(card.cardInformation())
+        print(card.description)
```
대부분 이런 부가 정보를 얻는 경우는 description을 생략해도 description을 가져와서 보여줍니다

godrm commented:
Card와 관련된 부분은 다음 단계를 진행하면서 개선해보세요.
